src/pdn/src/grid.cpp

```diff
@@ -0,0 +1,1114 @@
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "grid.h"
+
+#include <boost/geometry.hpp>
+
+#include "connect.h"
+#include "domain.h"
+#include "odb/db.h"
+#include "odb/dbTransform.h"
+#include "rings.h"
+#include "straps.h"
+#include "utl/Logger.h"
+
+namespace pdn {
+
+namespace bgi = boost::geometry::index;
+
+Grid::Grid(VoltageDomain* domain,
+           const std::string& name,
+           bool starts_with_power)
+    : domain_(domain),
+      name_(name),
+      starts_with_power_(starts_with_power),
+      allow_repair_channels_(false)
+{
+}
+
+Grid::~Grid()
+{
+}
+
+odb::dbBlock* Grid::getBlock() const
+{
+  return domain_->getBlock();
+}
+
+utl::Logger* Grid::getLogger() const
+{
+  return domain_->getLogger();
+}
+
+std::vector<odb::dbNet*> Grid::getNets(bool starts_with_power) const
+{
+  return domain_->getNets(starts_with_power);
+}
+
+const std::string Grid::typeToString(Type type)
+{
+  switch (type) {
+    case Core:
+      return "Core";
+    case Instance:
+      return "Instance";
+    case Existing:
+      return "Existing";
+  }
+
+  return "Unknown";
+}
+
+void Grid::addRing(std::unique_ptr<Rings> ring)
+{
+  if (ring != nullptr) {
+    ring->setGrid(this);
+    ring->checkLayerSpecifications();
+    rings_.push_back(std::move(ring));
+  }
+}
+
+void Grid::addStrap(std::unique_ptr<Straps> strap)
+{
+  if (strap != nullptr) {
+    strap->setGrid(this);
+    strap->checkLayerSpecifications();
+    straps_.push_back(std::move(strap));
+  }
+}
+
+void Grid::addConnect(std::unique_ptr<Connect> connect)
+{
+  if (connect != nullptr) {
+    for (const auto& conn : connect_) {
+      if (conn->getLowerLayer() == connect->getLowerLayer()
+          && conn->getUpperLayer() == connect->getUpperLayer()) {
+        getLogger()->error(
```

Comment:
Should we check for non-overlapping?  If you have a m1-m5 connect and get a m4-m5 connect that would seems like a conflict.

Reply:
Is it even legal to have a connect for non-adjacent layers?  I don't see a check for that.

Reply:
If you have a m1-m5 connect, you may still want to be able to connect to m4-m5, so that combination would be completely legal.
What do you mean by non-adjacent layers?

Reply:
Ok they can have different specification for m1-m5 and m4-m5 even though the former implicitly includes the later.